### PR TITLE
add fd approximants for td ones 

### PR DESCRIPTION
**Pull request by ahnitz**
This is still a work in progress.

### CODE REVIEW

**Comment by spxiwh** [comment]
I'll be on vacation until Monday, so if you want a review before then, please assign someone else.

**Comment by ahnitz** [comment]
@spxiwh We'll, this one isn't as time critical to me.

File: pycbc/waveform/waveform.py [hunk 1]
```diff
@@ -755,16 +820,24 @@ def imrphenomd_length_in_time(**kwds):
...
         cpu_fd[apx_int] = get_interpolated_fd_waveform
         _filter_time_lengths[apx_int] = _filter_time_lengths[apx]
 
         # We can also make a td version of this
-        apx_int = apx + '_FD'
-        cpu_td[apx_int] = get_td_waveform_from_fd
-        _filter_time_lengths[apx_int] = _filter_time_lengths[apx]
+        # This will override any existing approximants with the same name
+        # (ex. IMRPhenomXX)
+        cpu_td[apx] = get_td_waveform_from_fd
```
Comment (spxiwh) [inline pycbc/waveform/waveform.py]: There's a potential problem here if someone decides to make their *TD* waveform available also in the frequency domain in lalsimulation!

I think my preference is to handle the Phenom cases specially, but it's not a strong preference.

File: pycbc/waveform/waveform.py [hunk 2]
```diff
@@ -503,6 +503,72 @@ def get_fd_waveform(template=None, **kwargs):
...
+    # determine the duration to use
+    full_duration = duration = get_waveform_filter_length_in_time(**params)
+    nparams = params.copy()
+
+    while full_duration < duration * 1.5:
+        full_duration = get_waveform_filter_length_in_time(**nparams)
+        nparams['f_lower'] -= 1
+
+    if 'f_fref' not in nparams:
+        nparams['f_ref'] = params['f_lower']
```
Comment (spxiwh) [inline pycbc/waveform/waveform.py]: Some waveforms may not support this (ie. fref not being equal to f_lower). But let's deal with that as the need arises. Maybe we can push on waveform folks, or add special exceptions.

File: pycbc/waveform/waveform.py [hunk 3]
```diff
@@ -503,6 +503,72 @@ def get_fd_waveform(template=None, **kwargs):
...
+
+    # We'll try to do the right thing and figure out what the frequency
+    # end is. Otherwise, we'll just assume 2048 Hz.
+    # (consider removing as we hopefully have better estimates for more
+    # approximants
+    try:
+        f_end = get_waveform_end_frequency(**params)
+        delta_t = (0.5 / pnutils.nearest_larger_binary_number(f_end))
```
Comment (spxiwh) [inline pycbc/waveform/waveform.py]: Isn't max frequency also an optional parameter? That should take precedence if given, no?

File: pycbc/waveform/waveform.py [hunk 4]
```diff
@@ -503,6 +503,72 @@ def get_fd_waveform(template=None, **kwargs):
...
+                         "domain".format(params['delta_f'], params['approximant']))
+
+    hp.resize(tsamples)
+    hc.resize(tsamples)
+
+    # apply the tapering, we will use a safety factor here to allow for
+    # somewhat innacurate duration difference estimation.
+    window = (full_duration - duration) * 0.8
+    hp = wfutils.td_taper(hp, hp.start_time, hp.start_time + window)
+    hc = wfutils.td_taper(hc, hc.start_time, hc.start_time + window)
```
Comment (spxiwh) [inline pycbc/waveform/waveform.py]: Am I right that you're only tapering the start here? What about SpinTaylor approximants, which terminate *abruptly* in the time-domain?

File: pycbc/waveform/waveform.py (hunk 1 above)
Comment (ahnitz) [reply pycbc/waveform/waveform.py]: Indeed. I think we cross that bridge when we come to it though.

File: pycbc/waveform/waveform.py (hunk 3 above)
Comment (ahnitz) [reply pycbc/waveform/waveform.py]: Yep, I'll add that in before merging.

File: pycbc/waveform/waveform.py (hunk 4 above)
Comment (ahnitz) [reply pycbc/waveform/waveform.py]: You are correct, nothing is done for that at the moment. Primarily as I'm not sure what the natural choice is.

**Pull request merged**